third_party/gtp_ovs/ovs-gtp-patches/2.15/0023-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,1034 @@
++                        tun_flags |= TUNNEL_QFI;
++                        break;
+ 
+ 		default:
+ 			OVS_NLERR(log, "Unknown IP tunnel attribute %d",
+@@ -944,9 +954,6 @@ static int __ip_tun_to_nlattr(struct sk_buff *skb,
+ 	if (output->tp_dst &&
+ 	    nla_put_be16(skb, OVS_TUNNEL_KEY_ATTR_TP_DST, output->tp_dst))
+ 		return -EMSGSIZE;
+-	if ((output->tun_flags & TUNNEL_OAM) &&
+-	    nla_put_flag(skb, OVS_TUNNEL_KEY_ATTR_OAM))
+-		return -EMSGSIZE;
+ 	if (swkey_tun_opts_len) {
+ 		if (output->tun_flags & TUNNEL_GENEVE_OPT &&
+ 		    nla_put(skb, OVS_TUNNEL_KEY_ATTR_GENEVE_OPTS,
+@@ -964,6 +971,9 @@ static int __ip_tun_to_nlattr(struct sk_buff *skb,
+ 				 swkey_tun_opts_len, tun_opts))
+ 			return -EMSGSIZE;
+ 	}
++        if (output->qfi &&
++            nla_put_u8(skb, OVS_TUNNEL_KEY_ATTR_QFI, output->qfi))
++                return -EMSGSIZE;
+ 
+ 	return 0;
+ }
+diff --git a/datapath/linux/compat/gtp.c b/datapath/linux/compat/gtp.c
+index 725db82..91d6174 100644
+--- a/datapath/linux/compat/gtp.c
++++ b/datapath/linux/compat/gtp.c
+@@ -84,6 +84,17 @@ static int check_header(struct sk_buff *skb, int len)
+ 	return 0;
+ }
+ 
++struct gtpu_ext_hdr n_hdr = {
++        .type = 0x85,
++};
++
++struct gtpu_ext_hdr_pdu_sc pdu_sc_hdr = {
++        .len = 1,
++        .pdu_type = 0x0, /* PDU_TYPE_DL_PDU_SESSION_INFORMATION */
++        .qfi = 5,
++                .next_type = 0,
++};
++
+ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 			unsigned int hdrlen, u8 gtp_version,
+ 			__be64 tid, u8 flags, u8 type)
+@@ -116,15 +127,50 @@ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 		if (unlikely(opts_len)) {
+ 			struct gtpu_metadata *opts = ip_tunnel_info_opts(&tun_dst->u.tun_info);
+ 			struct gtp1_header *gtp1 = (struct gtp1_header *)(skb->data + sizeof(struct udphdr));
+-
+-			opts->ver = GTP_METADATA_V1;
+-			opts->flags = gtp1->flags;
+-			opts->type = gtp1->type;
++			struct gtpu_ext_hdr *geh;
++			geh = (struct gtpu_ext_hdr *) (gtp1 + 1);
++			n_hdr.type = geh->type;
++			struct gtpu_ext_hdr_pdu_sc *pdu_sc_hd;
++			pdu_sc_hd = (struct gtpu_ext_hdr_pdu_sc *) (geh + 1);
++			pdu_sc_hdr.qfi = pdu_sc_hd->qfi;
++			if (pdu_sc_hd->qfi) {
++				opts->ver = GTP_METADATA_EXT_HDR_DATA;
++				opts->flags = gtp1->flags;
++				opts->type = gtp1->type;
++				memcpy(opts->data, gtp1 + 1, sizeof(struct gtpu_ext_hdr) + sizeof( struct gtpu_ext_hdr_pdu_sc));
++				opts->header_data_len = sizeof(struct gtpu_ext_hdr) + sizeof(struct gtpu_ext_hdr_pdu_sc);
++				opts_len = opts_len + opts->header_data_len;
++			}
++			else {
++				opts->ver = GTP_METADATA_V1;
++				opts->flags = gtp1->flags;
++				opts->type = gtp1->type;
++			}
+ 			netdev_dbg(gtp->dev, "recved control pkt: flag %x type: %d\n", opts->flags, opts->type);
+ 			tun_dst->u.tun_info.key.tun_flags |= TUNNEL_GTPU_OPT;
+ 			tun_dst->u.tun_info.options_len = opts_len;
+ 			skb->protocol = 0xffff;         // Unknown
+ 		}
++		else {
```

Comment:
This is non needed. you need to check if here is extension hdr present and set opts_len currectly.